numpy/core/src/private/get_attr_string.h

```diff
@@ -1,37 +1,45 @@
 #ifndef __GET_ATTR_STRING_H
 #define __GET_ATTR_STRING_H
 
-static NPY_INLINE int
-_is_basic_python_type(PyObject * obj)
+static NPY_INLINE npy_bool
+_is_basic_python_type(PyTypeObject *tp)
 {
-    if (obj == Py_None ||
-            PyBool_Check(obj) ||
-            /* Basic number types */
+    return (
+        /* Basic number types */
+        tp == &PyBool_Type ||
 #if !defined(NPY_PY3K)
-            PyInt_CheckExact(obj) ||
-            PyString_CheckExact(obj) ||
+        tp == &PyInt_Type ||
 #endif
-            PyLong_CheckExact(obj) ||
-            PyFloat_CheckExact(obj) ||
-            PyComplex_CheckExact(obj) ||
-            /* Basic sequence types */
-            PyList_CheckExact(obj) ||
-            PyTuple_CheckExact(obj) ||
-            PyDict_CheckExact(obj) ||
-            PyAnySet_CheckExact(obj) ||
-            PyUnicode_CheckExact(obj) ||
-            PyBytes_CheckExact(obj) ||
-            PySlice_Check(obj)) {
+        tp == &PyLong_Type ||
+        tp == &PyFloat_Type ||
+        tp == &PyComplex_Type ||
 
-        return 1;
-    }
+        /* Basic sequence types */
+        tp == &PyList_Type ||
+        tp == &PyTuple_Type ||
+        tp == &PyDict_Type ||
+        tp == &PySet_Type ||
+        tp == &PyFrozenSet_Type ||
+        tp == &PyUnicode_Type ||
+        tp == &PyBytes_Type ||
+#if !defined(NPY_PY3K)
+        tp == &PyString_Type ||
+#endif
+
+        /* other builtins */
+        tp == &PySlice_Type ||
+        tp == Py_TYPE(Py_None) ||
+        tp == Py_TYPE(Py_Ellipsis) ||
+        tp == Py_TYPE(Py_NotImplemented) ||
+
+        /* TODO: ndarray, but we can't see PyArray_Type here */
 
-    return 0;
+        /* sentinel to swallow trailing || */
+        NPY_FALSE
```

Comment:
though that is very very minor as GCC just does all comparisons at once anyway and follows it up with some form of jump table.